iree_tests/README.md

```diff
@@ -0,0 +1,111 @@
+# IREE Tests
+
+This directory contains generated test suites for running through IREE's
+compiler and runtime tools.
+
+Each test suite has one folder per test program containing a few files:
+
+```
+[program name 1]/
+  model.mlir
+  input_0.npy
+  output_0.npy
+  test_data_flags.txt
+```
+
+Where:
+
+* `model.mlir` is in a format that is ready for use with `iree-compile`
+  (e.g. torch-mlir, stablehlo, tosa, linalg)
+* `input_0.npy` and `output_0.npy` files correspond to any number of program
+  inputs and outputs for one test case
+* `test_data_flags.txt` is a flagfile for use with
+  `iree-run-module --flagfile=test_data_flags.txt` of the format:
+
+  ```
+  --input=@input_0.npy
+  --expected_output=@output_0.npy
+  ```
+
+## Running the test suite
+
+Developer quickstart:
+
+```bash
+python -m pip install iree-compiler iree-runtime
+python -m pip install -r ./iree_tests/onnx/requirements.txt
+
+python ./iree_tests/onnx/import_tests.py
+python ./iree_tests/compile_tests.py
+python ./iree_tests/run_tests.py
+```
+
+Testing follows several stages:
+
+1. Import
+2. Compile
+3. Run
```

Comment:
<sub>(commenting here for threading)</sub>

@kumardeepakamd : 

> 1. Since we want this approach to scale, I am thinking not checking in the generated file but giving ability to generate the test and test output of a particular kind: onnx, torch mlir or linalg may work better. Similar approach can be applied to tests gotten from other frameworks, this also keeps the tests current. The IREE developer may want to start from torch MLIR, but a front end developer adding support for ONNX OP in torch MLIR or ONNX importer will like to have onnx test. Also, an end to end onnx test that compares the run result in ONNX runtime vs iree-run-module will be very useful.

---

@stellaraccident :

> At least for iree, we want test input artifacts snapshotted somewhere vs generated on every run. The CI volume and variants that need to check this stuff is too great to not have that step.
>
> But that can be a bespoke place that gets updated with "cooked" ready to test inputs materialized.
>
> Would be nice if this was not completely different tooling flows like it is today (versus more like priming the cache and then saving it -- same flow just starting in the middle for backend CI).

---

+1 to Stella's comments. For CI and downstream development flows, we need to decouple from the frontend framework. The full e2e test flows are useful for a certain segment of developers and users, but inserting a snapshot layer like this will keep complexity down.

I'm imagining that
* IREE will consume input MLIR, inputs, and outputs - _not_ ONNX models or Python sources. IREE will use a pinned version of the test suite repo that may lag behind the latest imported files.
  * This is currently just using checked in .mlir files, but we could also extend to download .mlir files from a cloud bucket if they grow beyond a certain size (beyond that git LFS can handle). The ONNX test suite has some code for "real" models generating inputs/outputs from a reference backend in Python. I want to avoid putting logic in the test suite for as long as possible.
  * For large tests, we can use IREE parameters too. Splats would save on space but we'd need to be careful about expected outputs / program behavior without real weights.
* The test repo itself could test the ONNX -> MLIR importer and update the generated files when appropriate. Similar story for JAX/StableHLO, TFLite/TOSA, and other programs in the test suite.

Reply:
Checking in .mlir for iree make sense. I guess, as needed that can be refreshed.